e2e/testcases/reconciler_finalizer_test.go

```diff
@@ -0,0 +1,685 @@
+	deployment2NN := types.NamespacedName{Name: "helloworld-2", Namespace: repoSyncNN.Namespace}
+	namespace1NN := types.NamespacedName{Name: repoSyncNN.Namespace}
+
+	nt.T.Cleanup(func() {
+		cleanupMultiLevel(nt,
+			rootSyncNN,
+			repoSyncNN,
+			deployment1NN,
+			deployment2NN,
+			namespace1NN)
+	})
+
+	// Namespace created for the RepoSync by test setup
+	namespace1 := rootRepo.Get(nomostest.StructuredNSPath(repoSyncNN.Namespace, repoSyncNN.Namespace))
+
+	// Add deployment-helloworld-1 to RootSync
+	deployment1Path := nomostest.StructuredNSPath(deployment1NN.Namespace, "deployment-helloworld-1")
+	deployment1 := loadDeployment(nt, "../testdata/deployment-helloworld.yaml")
+	deployment1.SetName(deployment1NN.Name)
+	deployment1.SetNamespace(deployment1NN.Namespace)
+	rootRepo.Add(deployment1Path, deployment1)
+	rootRepo.CommitAndPush("Adding deployment helloworld-1 to RootSync")
+	nt.WaitForRepoSyncs()
+	nomostest.WaitForCurrentStatus(nt, kinds.Deployment(), deployment1.GetName(), deployment1.GetNamespace())
+
+	// Add deployment-helloworld-2 to RepoSync
+	deployment2Path := nomostest.StructuredNSPath(deployment1NN.Namespace, "deployment-helloworld-2")
+	deployment2 := loadDeployment(nt, "../testdata/deployment-helloworld.yaml")
+	deployment2.SetName(deployment2NN.Name)
+	deployment2.SetNamespace(deployment1NN.Namespace)
+	nsRepo.Add(deployment2Path, deployment2)
+	nsRepo.CommitAndPush("Adding deployment helloworld-2 to RepoSync")
+	nt.WaitForRepoSyncs()
+	nomostest.WaitForCurrentStatus(nt, kinds.Deployment(), deployment2.GetName(), deployment2.GetNamespace())
+
+	// Tail reconciler logs and print if there's an error.
+	// This is necessary because if the RootSync/RepoSync are deleted, the
+	// normal Cleanup won't be able to find the reconcilers.
+	// Start here to catch both the finalizer injection and deletion behavior.
+	ctx, cancel := context.WithCancel(context.Background())
+	defer cancel()
+	go tailReconcilerLogs(ctx, nt, rootReconcilerObjectKey(rootSyncNN.Name))
+	go tailReconcilerLogs(ctx, nt, nsReconcilerObjectKey(repoSyncNN.Namespace, repoSyncNN.Name))
+
+	nt.T.Log("Enabling RootSync deletion propagation")
+	rootSync := fake.RootSyncObjectV1Beta1(rootSyncNN.Name)
+	err := nt.Get(rootSync.Name, rootSync.Namespace, rootSync)
+	if err != nil {
+		nt.T.Fatal(err)
+	}
+	if setDeletionPropagationPolicy(rootSync, metadata.DeletionPropagationPolicyForeground) {
+		err = nt.Update(rootSync)
+		if err != nil {
+			nt.T.Fatal(err)
+		}
+	}
+	nomostest.WatchForObject(nt, kinds.RootSyncV1Beta1(), rootSync.GetName(), rootSync.GetNamespace(), []nomostest.Predicate{
+		nomostest.StatusEquals(nt, kstatus.CurrentStatus),
+		nomostest.HasFinalizer(metadata.ReconcilerFinalizer),
+	})
+
+	nt.T.Log("Enabling RepoSync deletion propagation")
+	repoSync := rootRepo.Get(repoSyncPath)
+	if setDeletionPropagationPolicy(repoSync, metadata.DeletionPropagationPolicyForeground) {
```

Comment:
Is there a reason you chose to set the annotation directly rather than setting it from the git repo? I was expecting for a multi level test that we might test setting the annotation in the repo (not that it should have a big impact on functionality).

Reply:
This DOES set it in the git repo. `setDeletionPropagationPolicy` only modifies the object in memory. If it returns true, that means the object changed, which triggers the subsequent git add and git push.

Reply:
This might look excessive, but I want to make sure it's enabled, whether it was previously enabled or not. In the follow up PR, I change the setup code so it's always enabled by default for the e2e tests, which will cause this code block to no-op, but until that merges it still needs to make the change.

Reply:
🤦 you're right, disregard